[Clang] Fix crash on malformed std::partial_ordering static members

codepankajpy opened the pull request:
This fixes a crash (segmentation fault) when the standard library implementation of comparison categories (like `std::partial_ordering`) is malformed.

Specifically, if the static members (like `equivalent`) are defined as a primitive type (e.g., `int`) instead of the comparison category type itself, the compiler previously attempted to process them incorrectly, leading to a crash.

This adds strict type checking in `ComparisonCategoryInfo::lookupValueInfo`. If the found static member does not match the record type, it is rejected safely, triggering a diagnostic error instead of a crash.

Fixes #170015
Fixes #56571

codepankajpy committed:
[Clang] Fix crash on malformed std::partial_ordering static members

This fixes a crash (segmentation fault) when the standard library implementation of comparison categories (like `std::partial_ordering`) is malformed.

Specifically, if the static members (like `equivalent`) are defined as a primitive type (e.g., `int`) instead of the comparison category type itself, the compiler previously attempted to process them incorrectly, leading to a crash.

This patch adds strict type checking in `ComparisonCategoryInfo::lookupValueInfo`. If the found static member does not match the record type, it is rejected safely, triggering a diagnostic error instead of a crash.

Fixes #170015

tbaederr commented:
1) Add "Fixes <link to issue>" to your PR description
2) You need to add a test as well.

eoan-ermine commented on clang/lib/AST/ComparisonCategories.cpp:
```diff
@@ -49,7 +49,7 @@ bool ComparisonCategoryInfo::ValueInfo::hasValidIntValue() const {
   // actually have one (and only one) field.
   const auto *Record = VD->getType()->getAsCXXRecordDecl();
-  if (Record->getNumFields() != 1 ||
+  if (!Record || Record->getNumFields() != 1 ||
```
Shouldn't it be `!Record && (...)` instead of `!Record || ...`?

tbaederr commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -602,6 +602,7 @@ Bug Fixes to C++ Support
 - Fixed an issue where templates prevented nested anonymous records from checking the deletion of special members. (#GH167217)
 - Fixed spurious diagnoses of certain nested lambda expressions. (#GH149121) (#GH156579)
 - Fix the result of ``__is_pointer_interconvertible_base_of`` when arguments are qualified and passed via template parameters. (#GH135273)
+- - Fixed a crash when standard comparison categories (e.g. ``std::partial_ordering``) are defined with incorrect static member types. (#GH170015)
```
```suggestion
- Fixed a crash when standard comparison categories (e.g. ``std::partial_ordering``) are defined with incorrect static member types. (#GH170015)
```

ojhunt reviewed (changes requested):
LGTM, just some super minor nits on style, and I'd just like a few more tests - you've added logic to be safe against `using partial_ordering = int`, etc so we should make sure that's tested.

codepankajpy committed:
[Clang] Update regression tests for GH170015

Moved the regression test to a dedicated file (clang/test/SemaCXX/PR172001.cpp) to cover all requested edge cases, including typedefs and forward declarations. Reverted changes to clang/test/SemaCXX/cxx2a-three-way-comparison.cpp.

codepankajpy committed:
removed extra hyphen

cor3ntin reviewed (approved):
LGTM modulo nit

codepankajpy committed:
added new line

codepankajpy committed:
new line fix

zwuis reviewed (commented):
Thanks!

Can you test if this patch fixes #56571?

codepankajpy committed:
Merge branch 'llvm:main' into main

codepankajpy committed:
renamed ValueDecl to VD and ValueDeclRecord to VDRecord

codepankajpy committed:
added strong ordering testcase for issue 56571

code-pankaj commented:
@cor3ntin @zwuis Thanks for the review!

I have renamed the variables as requested. I also verified that this PR fixes #56571 and added a specific test case to cover it.

I have updated the PR description accordingly. Please let me know if there is anything else I can help with.

codepankajpy committed:
Fix clang-format violations

codepankajpy committed:
Update release notes to include GH56571

codepankajpy committed:
Merge branch 'main' into main

cor3ntin reviewed (approved):
LGTM - Will you need us to merge that for you? Thanks!
